Fix testcases and samples removing build_only tag

nashif opened the pull request:
We should use build_only for tests that are not supposed to run on anything,
not even Qemu. Those build_only tests are just compile test to make sure some
code does not bitrot. Everything else that does not have this tag should be
runnable on some hardware or simulation.

To identify what sanitycheck should run in Qemu, we add depends_on where it
makes sense or whitelist some samples for specific platforms. If a sample or a
test can run in Qemu but requires some interaction or evaluation of the console
we use the harness keyword to specify what sanitycheck should do, for example
we have a console harness for parsing console output and a keyboard harness
that can interact with the test using keyboard commands, for example for shell
tests. Those harnesses are not implemented yet in sanitycheck and are WIP, so
we use them now to exclude any tests or samples from running in sanitycheck.
Beside the above harnesses, we also have other setups that are more involved,
for example for sensors, networking and bluetooth.

Note that i started with commits per sample but that was too much, some commits
touch more than one sample/test.

Fixes #4724

nashif committed:
samples: pwm: depend on boards supporting PWM

Still need a better way to filter and a better way to support multiple
platforms with these tests, but for now we remove the build_only tag and
make it depend on real HW.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: move aio sample to tests

This is clearly a test and not a sample.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
build: remove stray CMakeLists.txt

This file is in the wrong place and does nothing.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: dma: use depends_on to whitelist platforms

Only build for those platforms that support DMA.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: remove standalone enc28j60 build test

We do already test building this driver somewhere else, this is a
duplicate.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: add a new harness keyword to schema

This keyword would mean that a special harness is needed to run the
tests sucessfully. This can be as simple as a loopback wiring or a
complete hardware test setup for sensor and IO testing. It is free form
initially and would be changed to be an enum once we have more values in
place.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: gpio: make test depend on gpio capability

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
boards: add supported features

Update board definitions with supported feature. This is temporary
until we have full coverage for device drivers in DTS, then we can use
DTS instead.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: i2c: make test depend on i2c

Build only for platforms that have an I2C driver.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: integrate nsim test in build test

This is just a build test, so make it part of the overall build test
instead of having a special test case for it.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: ipm: fix testcase tags

Add correct tag to the testcase.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: pci: fix tags

Add correct tag to the testcase.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: pinmux: depend on pinmux and support harness

Make sure we only build and run on supported platforms.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: pwm: fix dependencies and use depends_on

We should only build and run on supported devices.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: integrate quark clock test in build test

This is just a build test, so integrate in the build_only test.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: rtc: fix dependencies

Depend on RTC driver being available on the device.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: uart: fixed filtering and make test global

Remove build_only and improved platform filtering.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: harness support

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: kernel: move build test out of kernel

This test has nothing related to kernel.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: apds9960: remove arc related configs

Also remove unused STDOUT_CONSOLE

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: apds9960: fix dependencies

depend on i2c and gpio drivers being available.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: bme280: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: bmg160: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: bmi150: fix dependencies and test cleanup

Remove ARC unrelated configs

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: fxas21002: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: fxos8700: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: hts221: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: magn_polling: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: bmi160: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: max30101: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: mcp9808: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: sx9500: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: TH02: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: TMP112: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: rtc: fix dependencies and test cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: watchdog: fix dependencies and sample cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: benchmark: rename default .conf file

Have the default config be prj.conf

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: move gdb_server to test_build

No need for this to be a standalone test.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: bluetooth and net samples need test setup

remove build_only and make them depend on test harness (bluetooth)

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: entropy: filter based on HW capabilities

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: crypto: remove build_only and sample cleanup

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: drivers: remove build_only flag

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: remove build_only from various samples

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: remove more build_only tags

Remove build_only and add harness type needed for the sample/test to
allow running with sanitycheck and on devices once we have harness
support.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: remove build_only option

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: gdb server is build_only

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: timer_api: remove build_only, exclude arches

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

mbolivar commented on boards/arm/nucleo_f103rb/nucleo_f103rb.yaml:
```diff
@@ -6,3 +6,5 @@ toolchain:
   - gccarmemb
 ram: 20
+supported:
+  - pwm
```
I guess this means the board supports PWM.

Is there any way we can get this from the DTS or the pin mux information, perhaps as future work? I've  been thinking about how the "supported features" lists in the boards documentation tends to bitrot and how to fix that, and this seems like yet another place that can get out of sync with the code.

galak replied:
We really should get this from DT, we need to get sanitycheck to be DT aware

nashif replied:
I agree and I am fine getting this from DTS once we have enough coverage of all drivers and all boards, right now this is not the case, so until then...

nashif replied:
> I've been thinking about how the "supported features" lists in the boards documentation tends to bitrot and how to fix that, and this seems like yet another place that can get out of sync with the code.

yeah, I was thinking about the same originally, i.e. the supported hw table in the board documentation could be generated from DTS.

mbolivar replied:
> yeah, I was thinking about the same originally, i.e. the supported hw table in the board documentation could be generated from DTS.

It should be doable with another Sphinx extension once the Python DTS parsing is in place.

andrewboie commented:
I like this idea of defining test harnesses per test. I can conceive of some test cases to verify the correctness of some of the obscure linker stuff we are doing for memory protection, which would require a special test harness to inspect the produced ELF binary for the presence/absence of certain symbols or arrangement of stuff in memory.

andrewboie commented on tests/drivers/aio/aio_basic_api/CMakeLists.txt:
```diff
@@ -1,5 +0,0 @@
-include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Commit message states "tests: rename AIO api test" but I don't see a rename, it looks like the aio_basic_api test is simply being deleted in this commit. Maybe I'm missing something?

nashif replied:
fixed

andrewboie commented on tests/drivers/i2c/i2c_api/testcase.yaml:
```diff
@@ -1,7 +1,6 @@
+    depends_on: i2c
+    tags: drivers i2c
+    harness: sensor
+    platform_whitelist: quark_se_c1000_devboard quark_d2000_crb quark_se_c1000_ss_devboard # FIXME
```
maybe amend the FIXME to state what needs to be changed and why the platform_whitelist currently remains.

andrewboie commented on tests/drivers/ipm/testcase.yaml:
```diff
@@ -1,4 +1,4 @@
 tests:
   test:
     filter: not CONFIG_SOC_QUARK_SE_C1000_SS
```
hmm, wonder why this filter is here?

andrewboie commented on scripts/sanitycheck:
```diff
@@ -872,7 +872,7 @@ class MakeGenerator:
         args.append("BOARD={}".format(ti.platform.name))
         args.extend(extra_args)
-        if (ti.platform.qemu_support and (not ti.build_only) and
+        if (ti.platform.qemu_support and (not ti.build_only)  and
```
stray whitespace

andrewboie commented on samples/sensor/fxas21002/sample.yaml:
```diff
@@ -1,8 +1,8 @@
-    platform_whitelist: hexiwear_k64
+    harness: sensor
     tags: samples sensor
+    platform_whitelist: hexiwear_k64
```
does the fxas21002 appear on any other boards?

nashif replied:
it can be on other boards and on daughter boards, but I left it whitelisted because of other issues, ultimate goal is to remove all the whitelisting. There are some issues still with device naming convention which makes it fail if I build it for other boards.

andrewboie commented on samples/sensor/bmi160/sample.yaml:
```diff
@@ -1,8 +1,7 @@
-    platform_whitelist: arduino_101_sss
+    harness: sensor
     tags: sensors
+    depends_on: spi gpio
```
curious about the filtering.
we are saying that this depends on SPI and GPIO, but how do we know a bmi160 sensor is present on the board to run the test?

nashif replied:
based on recent developments related to DTS that are in the tree now, the sensors on a board will be defined in the board DTS, otherwise, if the sensor is off board, it will be matches based on a harness with that sensor, right now the harness is generic and called sensor.

andrewboie commented on samples/sensor/mcp9808/sample.yaml:
```diff
@@ -1,8 +1,7 @@
 tests:
   test:
-    build_only: true
-    platform_whitelist: frdm_k64f quark_d2000_crb
```
im a littlle confused on why for some tests we get rid of the platform_whitelist, but let it remain for others.

nashif replied:
see comment above.

andrewboie commented on tests/kernel/timer/timer_api/testcase.yaml:
```diff
@@ -2,7 +2,6 @@ tests:
-    build_only: true
     extra_args: CONF_FILE="prj_tickless.conf"
-    filter: CONFIG_BOARD_QEMU_X86
+    arch_exclude: riscv32 nios2
```
Maybe we need a feature flag indicating that tickless kernel is supported

nashif committed:
tests: rename AIO api test

Too much duplication in test structure, just call this api and being
under aio, explains that it is an API test for AIO.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
